SoC-2019-Microprojects.md

```diff
@@ -218,6 +218,28 @@ that we have `struct dir_iterator` (see dir-iterator.h), convert these
 to use the dir-iterator to simplify the code. Do only one conversion
 per microproject.
 
+### Let `git fetch <remote> HEAD` use the current branch name
+
+When fetching from a remote, one can specify which ref(s) to fetch.
+There is no convenient way to specify the current branch, though, and
+`HEAD` may be a good shortcut. Or `@`. Or `-`.
```

Comment:
I think `git fetch <remote> HEAD` already works and it might be difficult to convince people that it is ok to make it do something else than what it is currently doing. Or I am missing something?

Reply:
If by "already works" you mean to say "does something that is utterly incomprehensible and leaves the user totally puzzled", then I absolutely agree.

If you don't mean that, then I respectfully rescind my agreement with you.

Reply:
> If by "already works" you mean to say "does something that is utterly incomprehensible and leaves the user totally puzzled", then I absolutely agree.

(I assume you know what `git fetch origin HEAD` is doing???)

Reply:
Anyway I think a micro-project should be something more straightforward than that. At least I think the description should include something like "this means replacing the current behavior". I can add that if you want.

Reply:
> I assume you know what `git fetch origin HEAD` is doing???

From a quick test, I'm guessing that it fetches the remote ref named `HEAD` into `FETCH_HEAD`, which is what I had expected. Did I miss something (very likely so actually ;-) )?

Reply:
> From a quick test, I'm guessing that it fetches the remote ref named `HEAD` into `FETCH_HEAD`, which is what I had expected.

Technically, you are correct. Practically, the way `HEAD` is updated in a remote repository is... unintuitive. For example, last time I checked, when I changed my default branch on GitHub, it did not change `HEAD`. Other hosters like repo.or.cz do not even *have* the concept of a default branch. Git itself does not have any concept of a remote repository's default branch at all, it always defaults to cloning `master`.

Hence my statement that it is totally unintuitive what this does (and confusing).

Reply:
> At least I think the description should include something like "this means replacing the current behavior". I can add that if you want.

Sure, but let's add something that describes this a bit better. If you say that we change existing behavior, it kind of implies that the current behavior is well-defined: but it is not.

Reply:
OK, I mostly agree with you, but I think the confusing part is the very notion of `HEAD` in a remote repository, not really what `git fetch HEAD` does in this situation. I'd find it confusing if `git fetch remote master` was referring to the *remote* branch master, and `git fetch remote HEAD` to the *local* ref `HEAD`: to me the ref argument of `git fetch` is always a remote ref. 

> Git itself does not have any concept of a remote repository's default branch at all, it always defaults to cloning master.

That one is not true, though: `git clone` does clone the branch pointed-to by the remote's `HEAD` (modulo a "bug" when multiple branches point to the same commit, fixed in 1.8.4.3).

`-` would be very confusing to me too, since it usually means "the branch that used to be checked out before this one", not "the currently checked-out one".

Using `@` would be the less confusing to me. Perhaps `git fetch --current` would be best actually: no possible confusion, and explicit about what it would do.

Reply:
> For example, last time I checked, when I changed my default branch on GitHub, it did not change `HEAD`.

Just tested out of curiosity, and it did work for me: `git ls-remote --symref` shows the branch set as "default branch" on GitHub.

Reply:
@moy ?

Reply:
I [already explained](https://github.com/git/git.github.io/pull/372#discussion_r268823051) why I thought `-` would be confusing and inconsistent with other commands. `git fetch @` would probably be OK, though I'm not sure about all the uses of the `@` shortcut in Git.

Personally, I wouldn't use this feature, I always fetch everything so `git fetch` without argument is perfect for me. I guess people working on huge repositories may want to fetch only the current branch if only to gain a bit of download time, but I suspect this would be a relatively rarely used feature, hence wouldn't need a short syntax. If so `git fetch --current` would be best IMHO: no ambiguity, and not that long to type (especially if you have smart completion, `--c<TAB>` is just 4 keys).

Reply:
> `git fetch --current` would be best IMHO: no ambiguity, and not that long to type (especially if you have smart completion, `--c<TAB>` is just 4 keys).

:+1:

Reply:
Or `--current-branch`...